src/wayland/handlers/pointer_constraints.rs

```diff
@@ -27,11 +27,78 @@ impl PointerConstraintsHandler for State {
 
     fn cursor_position_hint(
         &mut self,
-        _surface: &WlSurface,
-        _pointer: &PointerHandle<Self>,
-        _location: Point<f64, Logical>,
+        surface: &WlSurface,
+        pointer: &PointerHandle<Self>,
+        location: Point<f64, Logical>,
     ) {
-        // TODO
+        if with_pointer_constraint(surface, pointer, |constraint| {
+            constraint.is_some_and(|c| c.is_active())
+        }) {
+            for (out, set) in &self.common.shell.read().workspaces.sets {
+                for workspace in &set.workspaces {
+                    // get the right window position based on if the window is fullscreen
+                    // this isn't the global position, rather the position inside the workspace
+                    let origin = if let Some(fullscreen) = workspace.get_fullscreen() {
+                        if fullscreen.wl_surface().as_deref() == Some(surface) {
```

Comment:
It's sadly not this easy. Any subsurface is also allowed to take a pointer lock, so we have to check the full surface tree and also take the subsurface offset into account.